src/server.rs

```diff
@@ -0,0 +1,512 @@
+            );
+
+        let delta = delta.collect();
+
+        self.0 = cids;
+
+        delta
+    }
+}
+
+impl<const S: usize, B> ServerBehaviour<S, B>
+where
+    B: Blockstore + Send + Sync + 'static,
+{
+    pub(crate) fn new(store: Arc<B>, protocol_prefix: Option<&str>) -> Self {
+        let protocol = stream_protocol(protocol_prefix, "/ipfs/bitswap/1.2.0")
+            .expect("prefix checked by beetswap::BehaviourBuilder::protocol_prefix");
+
+        ServerBehaviour {
+            protocol,
+            store,
+            peers_waitlists: FnvHashMap::default(),
+            global_waitlist: FnvHashMap::default(),
+            blockstore_tasks: Default::default(),
+            blockstore_tasks_abort_handles: FnvHashMap::default(),
+            outgoing_queue: Default::default(),
+            outgoing_event_queue: Default::default(),
+        }
+    }
+
+    fn schedule_store_get(&mut self, peer: PeerId, cid: CidGeneric<S>) {
+        let store = self.store.clone();
+        let (handle, reg) = AbortHandle::new_pair();
+
+        self.blockstore_tasks.push(
+            async move {
+                match Abortable::new(store.get(&cid), reg).await {
+                    Ok(result) => BlockstoreResult::Get(peer, cid, result),
+                    Err(_) => BlockstoreResult::Cancelled,
+                }
+            }
+            .boxed(),
+        );
+
+        self.blockstore_tasks_abort_handles
+            .insert((peer, cid), handle);
+    }
+
+    fn cancel_request(&mut self, peer: PeerId, cid: CidGeneric<S>) {
+        // remove pending blockstore read, if any
+        if let Some(abort_handle) = self.blockstore_tasks_abort_handles.remove(&(peer, cid)) {
+            abort_handle.abort();
+        }
+
+        // remove peer from the waitlist for cid, in case we happen to get it later
+        if let Entry::Occupied(mut entry) = self.global_waitlist.entry(cid) {
+            if entry.get().as_ref() == [peer] {
+                entry.remove();
+            } else {
+                let peers = entry.get_mut();
+                if let Some(index) = peers.iter().position(|p| *p == peer) {
+                    peers.swap_remove(index);
+                }
+            }
+        }
+
+        if let Some(peer_state) = self.peers_waitlists.get_mut(&peer) {
+            peer_state.0.remove(&cid);
+        }
+    }
+
+    pub(crate) fn process_incoming_message(&mut self, peer: PeerId, msg: ServerMessage) {
+        let Some(wantlist) = self.peers_waitlists.get_mut(&peer) else {
+            return; // entry should have been created in `new_connection_handler`
+        };
+
+        let wantlist_changes = wantlist.process_wantlist(msg.wantlist);
+
+        debug!("updating local wantlist for {peer}: {wantlist_changes:?}");
+
+        for change in wantlist_changes {
+            match change {
+                WantlistChange::Want(cid) => {
+                    self.schedule_store_get(peer, cid);
```

Comment:
Since this creates a new task each time, I think it would be more efficient if you request multiple CIDs at once.

Reply:
Switched to requesting all the CIDs from single wishlist update at once. Cancel handle logic got touch more complicated because of this change, still need to update that.

Reply:
Ditched cancellation, peer probably would not have enough time to send cancel anyway.